[Fix] Upgraded AWS go-sdk to V2 (only for AWS Canary Token Verification)

abmussani opened the pull request:
### Description:
The AWS SDK for Go ([aws-sdk-go](github.com/aws/aws-sdk-go)) has entered maintenance mode, as noted in its [README](https://github.com/aws/aws-sdk-go?tab=readme-ov-file#warning-this-sdk-is-in-maintenance-mode). This PR upgrades the SDK to v2, specifically for verifying AWS Canary tokens.

This upgrade addresses an issue where the previous version intermittently returned signature errors, incorrectly marking valid credentials as unverified.

Note: This upgrade is limited to AWS Canary token verification. The rest of the codebase will continue using the older AWS SDK version.

### Checklist:
* [ ] Tests passing (`make test-community`)?
* [x] Lint passing (`make lint` this requires [golangci-lint](https://golangci-lint.run/welcome/install/#local-installation))?

abmussani committed:
upgrade go-aws-sdk to v2 as the v1 is about to be deprecated and also returning errors randomly when verify canary tokens.
ran go mod tidy.

abmussani committed:
accidently removed the old version. readding old version

kashifkhan0771 reviewed (approved):
🚀

rgmz commented:
> This upgrade addresses an issue where the previous version intermittently returned signature errors, incorrectly marking valid credentials as unverified.

What part of the change fixes this issue?

abmussani commented:
> > This upgrade addresses an issue where the previous version intermittently returned signature errors, incorrectly marking valid credentials as unverified.
> 
> What part of the change fixes this issue?

Just migration of aws sdk to v2 and some code in canary.go file

abmussani commented:
> > Just migration of aws sdk to v2 and some code in canary.go file
> 
> That was never an issue with the canary flow AFAIk. It was in the normal Access Key / Session Key detector flows.
> 
> https://github.com/trufflesecurity/trufflehog/blob/f2dc96e8548d5fab0f6ffb3d47b9918b49fdde0a/pkg/detectors/aws/access_keys/accesskey.go#L280-L284

Some of the customer reported that Live AWS Canary tokens are not being verified. On Debugging, I noted that the old version of SDK was randomly causing signature verification error which got fixed when migrated to V2.

abmussani commented:
thank you @rgmz  for pointing out the usage in access_key.go I should also migrate the code to V2.

abmussani committed:
use aws go sdk v2 for access key verification

abmussani committed:
remove unused param

abmussani committed:
updated comments.

abmussani committed:
use verification client for better logging.
SDK has changed the error if the provided Id is invalid, incorporated the handling for that error as well.

abmussani commented:
@zricethezav I also request you to re-review the updated changes. TIA.

rgmz commented on pkg/detectors/aws/access_keys/accesskey.go:
```diff
@@ -198,118 +204,37 @@ const (
-		return false, nil, fmt.Errorf("request returned status %d with an unexpected reason (%s: %s)", res.StatusCode, body.Error.Code, body.Error.Message)
-	} else {
-		return false, nil, fmt.Errorf("request to %v returned unexpected status %d", res.Request.URL, res.StatusCode)
+		return false, nil, fmt.Errorf("request returned unexpected error: %s", err.Error())
```
```suggestion
		return false, nil, fmt.Errorf("request returned unexpected error: %w", err)
```

rgmz commented on pkg/detectors/aws/access_keys/accesskey.go:
```diff
@@ -198,118 +204,37 @@ const (
-
-	signature2 := aws.GetHMAC(hash, []byte(stringToSign)) // Get Signature HMAC SHA256
-	signature := hex.EncodeToString(signature2)
-
-	// TASK 4: ADD SIGNING INFORMATION TO THE REQUEST.
-	params.Add("X-Amz-Signature", signature)
-	req.Header.Add("Content-type", "application/x-www-form-urlencoded; charset=utf-8")
-	req.URL.RawQuery = params.Encode()
-
-	client := s.verificationClient
-	if client == nil {
-		client = defaultVerificationClient
-	}
+	// Create STS client
+	stsClient := sts.NewFromConfig(cfg)
```
> > This upgrade addresses an issue where the previous version intermittently returned signature errors, incorrectly marking valid credentials as unverified.

This does not resolve the mentioned issue. I can easily replicate the intermittent behaviour by taking a valid id+secret pair and adding a secondary invalid secret.
```
STORAGE_ACCESS_KEY_ID=AKIA...
#STORAGE_ACCESS_KEY_SECRET=...
STORAGE_ACCESS_KEY_SECRET=...
```

Output:
```sh
$ ./trufflehog filesystem /tmp/aws.txt
🐷🔑🐷  TruffleHog. Unearth your secrets. 🐷🔑🐷

✅ Found verified result 🐷🔑
Detector Type: AWS
Decoder Type: PLAIN
...

$ ./trufflehog filesystem /tmp/aws.txt
🐷🔑🐷  TruffleHog. Unearth your secrets. 🐷🔑🐷

Found unverified result 🐷🔑❓
Detector Type: AWS
Decoder Type: PLAIN
...
```

rgmz replied:
> But things work as expected if default AWS SDK client is being used.

By this do you mean omitting `config.WithHTTPClient(s.getClient()),`?

abmussani replied:
Correct. 

Yesterday, I was going through the AWS GO SDK documentation. They have provided a recommended way to customize the AWS HTTP client. [Link](https://docs.aws.amazon.com/sdk-for-go/v2/developer-guide/configure-http.html). Few of the properties are allowed to be modified. I am trying that out.

abmussani committed:
resolved the http client issue using with aws sdk.
passing the user-agent via middleware in aws sdk.
remove the hash logic since it was making the AWS Access key test as flaky by randomly generating the incorrect values.

abmussani commented:
Hey @rgmz, I have updated the PR and checked the issue you mentioned. The issue is no more reproducible with the changes. 

What's changed:
Since, `SaneHttpClient` is not compatible with AWS go SDK. I used `AWS BuildableClient` to send the requests. Along with that Middleware is added to inject the `TruffleHog + Suffix` in `User-Agent` Header.

Input:
```
STORAGE_ACCESS_KEY_ID=
#STORAGE_ACCESS_KEY_SECRET=...
STORAGE_ACCESS_KEY_SECRET=...
```
Output - First attempt:
```
$ go run . filesystem ~/Desktop/Trufflehog/aws/canary/accesskey.txt 
🐷🔑🐷  TruffleHog. Unearth your secrets. 🐷🔑🐷

2025-02-20T20:02:39+05:00	info-0	trufflehog	running source	{"source_manager_worker_id": "8FFgK", "with_units": true}
✅ Found verified result 🐷🔑
Detector Type: AWS
Decoder Type: PLAIN
...
```
Output - Second attempt:
```
$ go run . filesystem ~/Desktop/Trufflehog/aws/canary/accesskey.txt 
🐷🔑🐷  TruffleHog. Unearth your secrets. 🐷🔑🐷

2025-02-20T20:05:51+05:00	info-0	trufflehog	running source	{"source_manager_worker_id": "xZmZ7", "with_units": true}
✅ Found verified result 🐷🔑
Detector Type: AWS
Decoder Type: PLAIN
...

```

Also, I modified the test as generating the hash was causing flaky output. Sometime, it was getting passed and sometime not.
```
$ go test -tags=detectors ./pkg/detectors/aws/access_keys/
ok  	github.com/trufflesecurity/trufflehog/v3/pkg/detectors/aws/access_keys	(cached)
```

rgmz commented:
> The issue is no more reproducible with the changes.
> 
> What's changed: Since, `SaneHttpClient` is not compatible with AWS go SDK. I used `AWS BuildableClient` to send the requests. Along with that Middleware is added to inject the `TruffleHog + Suffix` in `User-Agent` Header.

The issue still exists, albeit it seems to be closer to a 1/10 chance than 1/2.

```sh
$ ./trufflehog filesystem /var/tmp/aws.txt
🐷🔑🐷  TruffleHog. Unearth your secrets. 🐷🔑🐷

2025-02-23T11:05:36-05:00       info-0  trufflehog      running source  {"source_manager_worker_id": "eCarM", "with_units": true}
2025-02-23T11:05:36-05:00       info-0  trufflehog.aws  Testing key...  {"detector_worker_id": "oWIUi", "detector": {"type":"AWS"}, "timeout": 10, "secret": "bad***"}
2025-02-23T11:05:36-05:00       info-0  trufflehog.aws  Testing key...  {"detector_worker_id": "oWIUi", "detector": {"type":"AWS"}, "timeout": 10, "secret": "37***Z"}
2025-02-23T11:05:36-05:00       info-0  trufflehog.aws  Testing key...  {"detector_worker_id": "oWIUi", "detector": {"type":"AWS"}, "timeout": 10, "secret": "bad***Y"}
2025-02-23T11:05:36-05:00       info-0  trufflehog.aws  Testing key...  {"detector_worker_id": "oWIUi", "detector": {"type":"AWS"}, "timeout": 10, "secret": "37***Z"}
Found unverified result 🐷🔑❓
Detector Type: AWS
Decoder Type: PLAIN
Raw result: AKIAJTQJNQSCYMNDT2OA
Resource_type: Access key
File: /var/tmp/aws.txt
Line: 1
```

rgmz commented:
> The issue is no more reproducible with the changes.
> 
> What's changed: Since, `SaneHttpClient` is not compatible with AWS go SDK. I used `AWS BuildableClient` to send the requests. Along with that Middleware is added to inject the `TruffleHog + Suffix` in `User-Agent` Header.

The issue still exists, albeit it seems to be closer to a 1/10 chance than 1/2.

```sh
$ ./trufflehog filesystem /var/tmp/aws.txt
🐷🔑🐷  TruffleHog. Unearth your secrets. 🐷🔑🐷

✅ Found verified result 🐷🔑
Detector Type: AWS
Decoder Type: PLAIN
Raw result: AKIA***
User_id: AIDA***
Arn: arn:aws:iam::***
Resource_type: Access key
Rotation_guide: https://howtorotate.com/docs/tutorials/aws/
Account: ***
File: /var/tmp/aws.txt
Line: 1

# ~10 attempts later

$ ./trufflehog filesystem /var/tmp/aws.txt
🐷🔑🐷  TruffleHog. Unearth your secrets. 🐷🔑🐷

Found unverified result 🐷🔑❓
Detector Type: AWS
Decoder Type: PLAIN
Raw result: AKIA***
Resource_type: Access key
File: /var/tmp/aws.txt
Line: 1
```

rgmz commented:
I think it's still worth retaining the 403 retry if there are multiple secrets for a given key. That _should_ catch most false negatives.
```diff
                        if verify && !isCanary {
-                               isVerified, extraData, verificationErr := s.verifyMatch(ctx, idMatch, secretMatch)
+                               isVerified, extraData, verificationErr := s.verifyMatch(ctx, idMatch, secretMatch, len(secretMatches) > 1)

...
-func (s scanner) verifyMatch(ctx context.Context, resIDMatch, resSecretMatch string) (bool, map[string]string, error) {
+func (s scanner) verifyMatch(ctx context.Context, resIDMatch, resSecretMatch string, retryOn403 bool) (bool, map[string]string, error) {
...
        resp, err := stsClient.GetCallerIdentity(ctx, &sts.GetCallerIdentityInput{})
        if err != nil {
                if strings.Contains(err.Error(), "StatusCode: 403") || strings.Contains(err.Error(), "InvalidClientTokenId") {
-                       return false, nil, nil
+                       if retryOn403 {
+                               return s.verifyMatch(ctx, resIDMatch, resSecretMatch, false)
+                       }
+                       return false, nil, err
```

rgmz commented:
Absolutely, it's down from >50% to 5-10%. Even so, it's worth retaining some comments — and retry handling — to account for this strange behaviour.

> I think it's still worth retaining the 403 retry if there are multiple secrets for a given key. That should catch most false negatives.

zricethezav commented:
> Absolutely, it's down from >50% to 5-10%. Even so, it's worth retaining some comments — and retry handling — to account for this strange behaviour.

Sounds like the current state of this PR is an improvement on what we have currently. I say let's merge this and follow up with an issue to track some of the remaining comments.

rgmz commented:
It is, but it's also removing some pretty important documentation and logic. I've already shared a reasonably implementation for logic, changing the comment (instead of removing it) wouldn't be that much extra work.

My fear is that this won't be followed up in a timely manner.

rgmz commented:
> I tried to put the retry logic back but the results remain same

From my testing it effectively eliminates false negatives.

abmussani committed:
re-added the retry logic if AWS returns 403

abmussani committed:
Merge branch 'main' into upgrade-aws-go-sdk-v2

rgmz commented:
Looks good!

abmussani committed:
Merge branch 'main' into upgrade-aws-go-sdk-v2
